### PR TITLE
[Solana]: Add or modify priority fee in the Solana raw transaction from a DApp

### PR DESCRIPTION
**Pull request by 10gic**
## Description

Currently, we only support updating the recent block hash in raw transactions from a DApp. In this PR, we add support for adding or modifying the priority fee in Solana raw transactions.

## How to test
Run Rust, C++, Kotlin, Swift tests

## Checklist

- [x] Create pull request as draft initially, unless its complete.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] If there is a related Issue, mention it in the description.

If you're adding a new blockchain

- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain.

### CODE REVIEW

File: rust/chains/tw_solana/src/modules/message_builder.rs [hunk 1]
```diff
@@ -71,7 +71,15 @@ impl<'a> MessageBuilder<'a> {
...
     pub fn build(self) -> SigningResult<VersionedMessage> {
         if let Some(ref raw_message) = self.input.raw_message {
-            return RawMessageBuilder::build(raw_message);
+            let mut result = RawMessageBuilder::build(raw_message);
+
+            if let Ok(version_message) = &mut result {
+                // Update the priority fee in raw message if priority fee is set in the signing input.
+                version_message
+                    .set_priority_fee(self.priority_fee_limit(), self.priority_fee_price());
+            }
```
Comment (satoshiotomakan) [inline rust/chains/tw_solana/src/modules/message_builder.rs:80]: We shouldn't modify the given raw message hiddenly, but rather expose an API for that, such as `SolanaTransaction.setComputeUnitLimit()`, `SolanaTransaction.getComputeUnitPrice()`, `SolanaTransaction.setComputeUnitPrice()`.

File: rust/chains/tw_solana/src/transaction/versioned.rs [hunk 2]
```diff
@@ -110,6 +114,74 @@ impl VersionedMessage {
...
+            ),
+            VersionedMessage::V0(v0) => {
+                (&mut v0.account_keys, &mut v0.header, &mut v0.instructions)
+            },
+        };
+
+        // Add compute budget account to the account keys if it doesn't exist.
+        if !account_keys.contains(&COMPUTE_BUDGET_ADDRESS) {
+            account_keys.push(*COMPUTE_BUDGET_ADDRESS);
+            header.num_readonly_unsigned_accounts += 1;
```
Comment (satoshiotomakan) [inline rust/chains/tw_solana/src/transaction/versioned.rs:142]: We should also take `address_lookup_table` into account. The lookup indices should be updated as well

**Review by satoshiotomakan — COMMENTED** [review]
Hi @10gic, thank you for the feature! I'm currently working on the same functionality, but in a more generic way with exposing an API. We can't just break backward compatibility and add a priority fee price hiddenly.
I'll open a PR this week with the feature

**Comment by 10gic** [comment]
@satoshiotomakan I greatly appreciate your feedback.

**Comment by satoshiotomakan** [comment]
@10gic I'm closing the PR because we decided to take a slightly different path of the implementation.
I'll reuse some parts of the code if you are ok

**Comment by 10gic** [comment]
> @10gic I'm closing the PR because we decided to take a slightly different path of the implementation. I'll reuse some parts of the code if you are ok

Of course, please feel free to use the code from the PR. It's my pleasure.

File: rust/chains/tw_solana/src/modules/message_builder.rs (hunk 1 above)
Comment (10gic) [reply rust/chains/tw_solana/src/modules/message_builder.rs:80]: Hi @satoshiotomakan, do you mean we should implement updating the priority fee just like the SolanaTransaction.updateBlockhashAndSign?

Initially, I had implemented it in a similar way just liking SolanaTransaction.updateBlockhashAndSign, but I abandoned that approach due to the following reasons:

Reason 1: the "warning" comments in the code. 
https://github.com/trustwallet/wallet-core/blob/58fdabf62c17149862552da32f5b56a0aea21dae/include/TrustWalletCore/TWSolanaTransaction.h#L19-L29

Reason 2: the return value of SolanaTransaction.updateBlockhashAndSign is SigningOutput. It's not good for case of getting signature from signature server. In order to use function compileWithSignatures, user need construct SingingInput from this SigningOutput. It's a bit cumbersome.

Reason 3: the approach in this PR is very convenient for the users.
```
    let input = Proto::SigningInput {
         raw_message: Some(decoded_tx),
         priority_fee_limit: Some(Proto::PriorityFeeLimit { limit: 10_000 }),
         priority_fee_price: Some(Proto::PriorityFeePrice { price: 1_000 }),
         ..Proto::SigningInput::default()
     };
```
BTW, I think the method in this PR is not a **hidden** approach, as the priority_fee_limit and priority_fee_price settings in SigningInput are explicitly set by the user

Comment (satoshiotomakan) [reply rust/chains/tw_solana/src/modules/message_builder.rs:80]: I totally agree on the `Reason 2`. The interface I plan to add will look like this:
* SolanaTransaction.getComputeUnitLimit(encodedTransaction: String) -> String?, where:
  * encodedTransaction - base64 encoded transaction
  * return - optional Unit Limit as a decimal string
* SolanaTransaction.setComputeUnitLimit(encodedTransaction: String, limit: String) -> String, where:
  * encodedTransaction - base64 encoded transaction
  * limit - decimal string
  * return - base64 encoded updated transaction
* SolanaTransaction.getComputeUnitPrice(encodedTransaction: String) -> String?, where
  * encodedTransaction - base64 encoded transaction
  * return - optional Unit Price as a decimal string
* SolanaTransaction.setComputeUnitPrice(encodedTransaction: String, price: String) -> String, where:
  * encodedTransaction - base64 encoded transaction
  * price - decimal string
  * return - base64 encoded updated transaction

So you'll get a serialized transaction back, and you can call `TransactionDecoder.decode()` to convert it to `Solana.Proto.RawMessage`

Comment (satoshiotomakan) [reply rust/chains/tw_solana/src/modules/message_builder.rs:80]: Regarding `Reason 3` - probably it's a matter of taste. Anyway in TrustWallet we need to be able to get the previous `ComputeUnitLimit` and `ComputeUnitPrice` to compare them with our own estimations.
That's why we decided to add dedicated FFI functions

**Pull request closed without merging**